### PR TITLE
Gabrieltabares/ch11288/unbound lookup forward address

### PR DESCRIPTION
**Pull request by gabrielexoscale**
This PR adds the possibility of dynamically setting forward hosts via lookup.

In order to do this, you add a address_lookup field with the hiera lookup and it will use that instead of the 'address' values.

The main reason for this change is to be able to set the consul servers as default forward servers.

It has been tested in preprod, in consul servers (which will not have this set) and other servers (which will).

### CODE REVIEW

**Review by irregulator — COMMENTED** [review]
> The main reason for this change is to be able to set the consul servers as default forward servers.

If I understand correctly the goal why don't we just change the `unbound::forward` keys in hiera where needed? Either per location (as it's now configured) or per other appropriate level in the hierarchy.

The `$address` variable is already looked-up from hiera as part of the `$forward` variable.

`exoscale::base`:
```
class { "unbound":
      forward => hiera('unbound::forward') + hiera('unbound::extra_forwards', {})
}
```

`unbound`:
```
if $forward {
    create_resources('unbound::forward', $forward)
}
```

**Comment by gabrielexoscale** [comment]
Because you cannot look up arrays in Hiera.

If I set the resource in hiera like this

```
unbound::forward:
  '.':
    address: '%{hiera("consul::servers")}'
```

what you get is a string with '["consul server 1", "consul server 2", "consul server 3"]'

There is no way to get anything other than a string interpolation as a lookup in hiera, so this gets around the issue by doing the lookup at the last minute.

File: manifests/forward.pp [hunk 1]
```diff
@@ -26,6 +26,7 @@
   $zone          = $name,
   $forward_first = 'no',
   $config_file   = $unbound::params::config_file,
+  $address_lookup = ''
```
Comment (irregulator) [inline manifests/forward.pp:34]: Can you please add a couple of doc lines regarding this newly introduced parameter?

**Comment by exo-cedric** [comment]
It would be nice to have a few-liners comment explaining what the new parameter does and the kind of value it expects (like has been done for existing parameters)
(if I did my home work and read https://puppet.com/docs/puppet/5.5/function.html#lookup correctly, it expect a String that matches a Hiera variable name, right ?)

**Review by irregulator — COMMENTED** [review]
okay it makes sense now :) given the above comment and https://github.com/exoscale/puppet/pull/3466.

**Comment by exo-cedric** [comment]
An alternative way that would preserve backward-compatilibity without adding a new parameter and handle both String and Array nicely is (in ERB template):
```
<% [@address].flatten.each do |addr| -%>
<% if addr[0..7] == 'hiera://' -%>
  forward-addr: <%= scope[addr[8..-1]] %>
<% else -%>
  forward-addr: <%= addr %>
<% end -%>
<% end -%>
```
[EDITED] Forgot the lookup

[EIDTED] Or maybe it's `scope.function_hiera` (as per https://stackoverflow.com/questions/27296607/erb-template-not-picking-up-variables-from-hieradata#27302861)

**Comment by gabrielexoscale** [comment]
@exo-cedric 

That wont' work. The issue is that you can't lookup anything but a string in hiera, so you get a string that may contain a single value or an array of values.

The only way to get a real array is to actually do the lookup at the last minute.

Also, I thought of overriding the address value but it's a magic substring and, to my eyes, less clear than adding a separate key.

**Review by wdullaer — APPROVED** [review]

File: manifests/forward.pp (hunk 1 above)
Comment (gabrielexoscale) [reply manifests/forward.pp:34]: Done.

File: templates/forward.erb [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
   forward-addr: <%= addr %>
 <% end -%>
-<% elsif @address != '' -%>
-  forward-addr: <%= @address %>
+<% elsif @forward_address != '' -%>
+  forward-addr: <%= @forward_address %>
 <% end -%>
 <% if @forward_first != 'no' -%>
   forward-first: <%= @forward_first %>
-<% end -%>
```
Comment (LaCap) [inline templates/forward.erb]: There is no linting here but the no new line would be seen as an error :)

File: templates/forward.erb [hunk 3]
```diff
@@ -1,12 +1,12 @@
 forward-zone:
   name: "<%= @zone %>"
-<% if @address.is_a? Array -%>
-<% @address.each do |addr| -%>
+<% if @forward_address.is_a? Array -%>
```
Comment (LaCap) [inline templates/forward.erb:3]: Wouldn't be easier to always pass an array?

Comment (LaCap) [reply templates/forward.erb:3]: Forget what I said... Just noticed the:
```
  else {
    $forward_address = $address
  }
```

**Review by LaCap — APPROVED** [review]

File: templates/forward.erb (hunk 2 above)
Comment (gabrielexoscale) [reply templates/forward.erb]: Fixed!

**Pull request merged**